sdk/src/wallet/utxo.ts

```diff
@@ -279,33 +310,51 @@ export async function estimateTxFee(
     // TODO: allow submitting the UTXOs, fee estimate and confirmed transactions
     // to avoid fetching them again.
     const esploraClient = new EsploraClient(addressInfo.network);
+    const ordinalsClient = new OrdinalsClient(addressInfo.network);
 
     let confirmedUtxos: UTXO[] = [];
+    let outputsFromAddress: OutputJson[] = [];
+
     if (feeRate) {
-        confirmedUtxos = await esploraClient.getAddressUtxos(fromAddress);
+        [confirmedUtxos, outputsFromAddress] = await Promise.all([
+            esploraClient.getAddressUtxos(fromAddress),
+            // cardinal = return UTXOs not containing inscriptions or runes
+            addressInfo.type === AddressType.p2tr ? ordinalsClient.getOutputsFromAddress(fromAddress, 'cardinal') : [],
+        ]);
     } else {
-        [confirmedUtxos, feeRate] = await Promise.all([
+        [confirmedUtxos, feeRate, outputsFromAddress] = await Promise.all([
             esploraClient.getAddressUtxos(fromAddress),
             esploraClient.getFeeEstimate(confirmationTarget),
+            // cardinal = return UTXOs not containing inscriptions or runes
+            addressInfo.type === AddressType.p2tr ? ordinalsClient.getOutputsFromAddress(fromAddress, 'cardinal') : [],
         ]);
     }
 
     if (confirmedUtxos.length === 0) {
         throw new Error('No confirmed UTXOs');
     }
 
-    const possibleInputs = await Promise.all(
+    const outpointsSet = new Set(outputsFromAddress.map((output) => output.outpoint));
+    const possibleInputs: Input[] = [];
+
+    await Promise.all(
         confirmedUtxos.map(async (utxo) => {
             const hex = await esploraClient.getTransactionHex(utxo.txid);
             const transaction = Transaction.fromRaw(Buffer.from(hex, 'hex'), { allowUnknownOutputs: true });
-
-            return getInputFromUtxoAndTx(
+            const input = getInputFromUtxoAndTx(
                 addressInfo.network as BitcoinNetworkName,
                 utxo,
                 transaction,
                 addressInfo.type,
                 publicKey
             );
+
+            // to support taproot addresses we want to exclude outputs which contain inscriptions
```

Comment:
Btw do you know if the Ord API will index the mempool txs appropriately? Seems to me there could be a race condition if esplora returns a UTXO from the mempool and Ord has not yet updated that - cc: @sander2

Reply:
The answers are [1](https://github.com/bob-collective/bob/pull/408#discussion_r1841929462) and [2](https://github.com/bob-collective/sats-wagmi/pull/125#discussion_r1842863707)